### ISSUE TITLE
Double-quotation marks aren't accepted for parametrised sessions

### ISSUE THREAD

**Issue by smithfred** [theacodes/nox#396]
Given the following:

```
@nox.parametrize("foo", ["bar", "baz"])
def test(...
```

I'd expect to be able to run the test with a given value by calling either `nox -s "test(foo='bar')"` or `nox -s 'test(foo="bar")'`, since single and double-quotation marks are equivalent in Python generally.

However, the latter form (or similar variations) fails.

Being able to use either is useful in terms of managing shell variable expansion etc.

**Comment by theacodes** [comment]
This should be pretty straightforward if we normalize the quotes for the `--session` arg.

**Comment by t-eckert** [comment]
@theacodes, I will take this one, but I will need some help. I have a fork and a branch. I started digging into the code for the first time. I _think_ what I need to do is modify code in the `tasks` module, but I could use some guidance to get on the right track.

Thank you!

**Comment by theacodes** [comment]
Yeah, I'm thinking that you'll need to normalize the user-provided session names either [here](https://github.com/theacodes/nox/blob/main/nox/tasks.py#L129) or [here](https://github.com/theacodes/nox/blob/main/nox/manifest.py#L110). I feel like the latter is probably the more logical location.

**Comment by t-eckert** [comment]
Ok, I think I will squeeze it into the line above where the session is appended to the `queue`. For "normalization"... I could write a function that checks if the string begins with a single quote then just swap every single to double or vice versa. I worry I would miss an edge case or change something unintentionally (a contraction perhaps (though unlikely)). 

Is there a more elegant way to do this normalization that you can think of?

If I have to write a helper function, where do I put it in the codebase?

**Comment by cjolowicz** [comment]
How would we handle the case where the string itself contains quotes?

TBH I think we're better off documenting that `--session` must be used with the session name as displayed by `nox --list`. `nox.parametrize` can take any Python value, so fundamentally Nox would have to do what Black does to normalize these.

**Comment by t-eckert** [comment]
I'll leave that discussion up to the maintainers. But since you mentioned Black, I wonder if there is a solution where Black's functionality could be leveraged to normalize the string.

**Comment by t-eckert** [comment]
I will hang in a holding pattern on this until I get some direction from the maintainers. Still willing to do the work.

**Comment by cjolowicz** [comment]
I'll provide some additional info, hoping that it helps us decide what to do here.

Here are some examples of valid session names:

```
* test(arg='jane')
* test(arg="jane's")
* test(arg='"hello world"')
* test(arg=[42])
* test(arg=datetime.datetime(1980, 1, 1, 0, 0))
* test(arg=<user_nox_module.Foo object at 0x7f85968b4640>)
```

<details>
<summary>To reproduce, run `nox -l` with the noxfile.py below:</summary>

```python
import datetime
import nox

class Foo:
    pass

@nox.session
@nox.parametrize(
    "arg",
    [
        "jane",
        "jane's",
        '"hello world"',
        datetime.datetime(1980, 1, 1),
        [42],
        Foo(),
    ],
)
def test(session, arg):
    pass
```
</details>

Replacing double quotes by single quotes will work for the first example, but break the second and third examples where the strings contain quotes.

The remaining examples demonstrate how hard it is to normalize session names in the general case:

- The list could have a trailing comma.
- The date construction might have different whitespace between arguments.
- An instance of a user-defined class could even contain a memory address and the internal name under which the Noxfile was imported.

We can normalize the first four examples using `ast.parse` and `ast.literal_eval`:

```python
>>> import ast
>>> repr(ast.literal_eval(ast.parse("\"jane's\"",mode="eval")))
'"jane\'s"'
```

The datetime example can be parsed, but will raise `ValueError` when evaluated because it is not a Python literal.

The user object cannot even be parsed, and will raise `SyntaxError`.

[ast.parse]: https://docs.python.org/3/library/ast.html#ast.parse
[ast.literal_eval]: https://docs.python.org/3/library/ast.html#ast.literal_eval

**Comment by cjolowicz** [comment]
If we do want to provide this kind of normalization, here's one thing we could try:

1. Parse the session name using [ast.parse].
2. On `SyntaxError`, give up and use the original session name.
3. If the AST does not have the expected form, give up and use the original session name. The expected form would be a function call (the session) using only keyword arguments (the parameters).
4. For each keyword argument, attempt to load the value using [ast.literal_eval].
5. On `ValueError`, use the original source for this value ([ast.get_source_segment]).
6. Otherwise, use `repr` to normalize the value.
7. Reassemble the session name using the normalized values, where available.

There is still a possibility that a session name is parsed into something that *looks* right and raises no exceptions, but is actually bogus. That's because `repr` could return pretty much anything, it's not required that you can `eval` it back into the original value.

[ast.parse]: https://docs.python.org/3/library/ast.html#ast.parse
[ast.literal_eval]: https://docs.python.org/3/library/ast.html#ast.literal_eval
[ast.get_source_segment]: https://docs.python.org/3/library/ast.html#ast.get_source_segment

**Comment by cjolowicz** [comment]
Fixed in #417
